File: tests/test_extractor.cpp
#include "extractor.cpp"
#include "extractor.hpp"

#include <gtest/gtest.h>
#include <sndfile.h>

#include <algorithm>
#include <complex>
#include <ctime>
#include <filesystem>
#include <iostream>

// Audio fixture
std::filesystem::path AUDIO_PATH =
    std::filesystem::path(__FILE__).parent_path() / "data" / "audio0.wav";

TEST(TestWav, WavIntoVector) {
    SF_INFO file_info;
    SNDFILE* file  = sf_open(AUDIO_PATH.string().c_str(), SFM_READ, &file_info);
    int samplerate = file_info.samplerate;
    if (sf_error(file) || samplerate < 0) {
        throw std::runtime_error(sf_strerror(file));
        sf_close(file);
    }
    std::vector<float> empty_data(250), audio_data(250);
    sf_read_float(file, audio_data.data(), audio_data.size());
    ASSERT_NE(audio_data, empty_data);
}

TEST(TestExtractor, GammatoneFilterBank) {
    // Example input
    int num_samples = 250;
    SF_INFO file_info;
    SNDFILE* file = sf_open(AUDIO_PATH.string().c_str(), SFM_READ, &file_info);
    std::vector<float> input(num_samples);
    sf_read_float(file, input.data(), input.size());

    // Construct extractor
    auto extractor = ao::extractor::GammatoneFilterbank<float>(
        /* num_samples */ num_samples,
        /* num_features */ 64,
        /* sample_rate */ file_info.samplerate,
        /* low_Hz */ 50,
        /* high_Hz */ 8000);
    std::cout << "Center frequencies: ";
    for (auto& filter : extractor.filters) {
        std::cout << filter.cf << " ";
    }
    std::cout << std::endl;

    // Execute extractor
    std::vector<float> output = extractor.compute(input);
    EXPECT_EQ(output, extractor.compute(input)); // Test second execution

    // Invalid input (too short)
    std::vector<float> invalid_input(num_samples-1);
    sf_read_float(file, invalid_input.data(), invalid_input.size());
    EXPECT_THROW(extractor.compute(invalid_input), std::invalid_argument);
}

/**
 * @brief Extractors overload the feature "compute" method with several
 * signatures. Test that all of them work and return the same values.
 *
 */
TEST(TestExtractor, ComputeOverloads) {
    // TODO parametrize input
    int num_samples = 250;
    SF_INFO file_info;
    SNDFILE* file = sf_open(AUDIO_PATH.string().c_str(), SFM_READ, &file_info);
    std::vector<float> input(num_samples);
    sf_read_float(file, input.data(), input.size());

    // TODO parametrize with different extractors
    auto extractor = ao::extractor::GammatoneFilterbank<float>(
        /* num_samples */ num_samples,
        /* num_features */ 64,
        /* sample_rate */ file_info.samplerate,
        /* low_Hz */ 50,
        /* high_Hz */ 8000);
    std::vector<float> output = extractor.compute(input);
    // operator()
    EXPECT_EQ(output, extractor(input));
}

File: tests/test_extractor.py
import ao

import math
import pytest
import warnings
import numpy as np

# TODO test overide Extractor


@pytest.mark.parametrize('frame_duration', [1, 10, 100, 1000])
@pytest.mark.parametrize('num_features', [64, 256, 0])
def test_extractor(audio_data, frame_duration, num_features):
    data, fs = audio_data
    num_samples = int(frame_duration / 1000 * fs)  # samples per frame
    extractor = ao.extractor.GammatoneFilterbank(num_samples, num_features, fs)
    for frame in ao.dataset.audio.frames(data, num_samples):
        # TODO do not average channels, use extractor on_channel
        output = extractor(frame.mean(axis=1))
        assert len(output) == num_features


@pytest.mark.parametrize('frame_duration', [100])
@pytest.mark.parametrize('num_features', [256])
def test_transform(audio_data, frame_duration, num_features):
    data, fs = audio_data
    num_samples = int(frame_duration / 1000 * fs)  # samples per frame
    no_transform = ao.extractor.GammatoneFilterbank(
        num_samples, num_features, fs, lambda x: x
        )
    transform_log10 = ao.extractor.GammatoneFilterbank(
        num_samples, num_features, fs, lambda x: math.log10(x)
        )
    for frame in ao.dataset.audio.frames(data, fs, frame_duration):
        raw = no_transform(frame.mean(axis=1))
        transformed = transform_log10(frame.mean(axis=1))
        assert all([r != t for r, t in zip(raw, transformed)])
        assert all(np.log10(raw) == transformed)


def test_gammatone_filterbank():
    with warnings.catch_warnings():
        warnings.simplefilter("error")
        extractor = ao.extractor.GammatoneFilterbank()
        for filter in extractor.filters:
            assert isinstance(filter.cf, float)
            assert isinstance(filter.gain, float)
            assert isinstance(filter.a, list)
